### PR TITLE
1.20.6

### PR DESCRIPTION
**Pull request by Mrbysco**
Initial update to 1.20.6. 

I was able to abstract/implement the network message classes into the Xplat source-set.

There's still some TODOs like figuring out what to do with the `nbt` keys from the item related json.

### CODE REVIEW

File: Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapedBookRecipe.java [hunk 1]
```diff
@@ -50,35 +51,41 @@ public RecipeSerializer<?> getSerializer() {
...
-						ExtraCodecs.strictOptionalField(ItemStack.ITEM_WITH_COUNT_CODEC, "result", ItemStack.EMPTY).forGetter(bookRecipe -> bookRecipe.result),
-						ExtraCodecs.strictOptionalField(ResourceLocation.CODEC, "book", null).forGetter(bookRecipe -> bookRecipe.outputBook)
+						ItemStack.STRICT_CODEC.optionalFieldOf("result", ItemStack.EMPTY).forGetter(bookRecipe -> bookRecipe.result),
+						ResourceLocation.CODEC.optionalFieldOf("book", null).forGetter(bookRecipe -> bookRecipe.outputBook)
 				)
 						.apply(instance, ShapedBookRecipe::new)
 		);
+		public static final StreamCodec<RegistryFriendlyByteBuf, ShapedBookRecipe> STREAM_CODEC = StreamCodec.of(
+				ShapedBookRecipe.Serializer::toNetwork, ShapedBookRecipe.Serializer::fromNetwork
+		);
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapedBookRecipe.java]: Please do a proper stream codec

File: Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapelessBookRecipe.java [hunk 2]
```diff
@@ -76,41 +77,43 @@ public static class Serializer implements RecipeSerializer<ShapelessBookRecipe>
...
 								)
 								.forGetter(bookRecipe -> bookRecipe.ingredients),
-						ExtraCodecs.strictOptionalField(ResourceLocation.CODEC, "book", null).forGetter(bookRecipe -> bookRecipe.outputBook)
+						ResourceLocation.CODEC.optionalFieldOf("book", null).forGetter(bookRecipe -> bookRecipe.outputBook)
 				)
 						.apply(instance, ShapelessBookRecipe::new)
 		);
+		public static final StreamCodec<RegistryFriendlyByteBuf, ShapelessBookRecipe> STREAM_CODEC = StreamCodec.of(
+				ShapelessBookRecipe.Serializer::toNetwork, ShapelessBookRecipe.Serializer::fromNetwork
+		);
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapelessBookRecipe.java]: again please use a proper stream codec instead of using manual encode and decode methods

File: Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java [hunk 3]
```diff
@@ -26,19 +30,22 @@
...
 
 	private ItemStackUtil() {}
 
-	public static Triple<ResourceLocation, Integer, CompoundTag> parseItemStackString(String res) {
-		String nbt = "";
-		int nbtStart = res.indexOf("{");
-		if (nbtStart > 0) {
-			nbt = res.substring(nbtStart).replaceAll("([^\\\\])'", "$1\"").replaceAll("\\\\'", "'");
-			res = res.substring(0, nbtStart);
+	public static Triple<ResourceLocation, Integer, DataComponentMap> parseItemStackString(String res) {
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java]: Maybe it's time for a purpose built record here

File: Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+@Mixin(Component.Serializer.class)
+public interface AccessorComponentSerializer {
+	@Invoker("serialize")
+	public static JsonElement invokeSerialize(Component component, HolderLookup.Provider provider) {
+		throw new AssertionError();
+	}
+}
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java]: Why is this needed? can't you use the codec with a RegistryOps<JsonElement>?

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

File: Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java (hunk 4 above)
Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java]: I could, but that would basically be an exact copy of an already existing method

File: Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java (hunk 3 above)
Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java]: Probably, I didn't want to change anything without talking it over first

File: Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapedBookRecipe.java (hunk 1 above)
Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapedBookRecipe.java]: Blame vanilla serializer, it was inspired by that.

Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/common/recipe/ShapedBookRecipe.java]: As discussed on Discord since mojang allows defining components in recipe outputs now the custom recipe types aren't really needed

**Comment by Mrbysco** [comment]
As discussed in Discord. The custom book recipe types have been removed since you can specify components in vanilla recipe outputs now.

File: Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java (hunk 4 above)
Comment (Minecraftschurli) [reply Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java]: still preferred to adding mixins

Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/mixin/AccessorComponentSerializer.java]: I've removed the mixin in favor of copying the method

File: Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java [hunk 5]
```diff
@@ -55,32 +62,34 @@ public static Triple<ResourceLocation, Integer, CompoundTag> parseItemStackStrin
...
-				tag = TagParser.parseTag(nbt);
+				ItemParser.ItemResult result = parser.parse(new StringReader(res.toString() + components)); //TODO: Check if the parsed result is correct
+				componentMap.addAll(result.components());
 			} catch (CommandSyntaxException e) {
 				throw new RuntimeException("Failed to parse ItemStack JSON", e);
 			}
 		}
 
-		return ImmutableTriple.of(key, countn, tag);
+		return ImmutableTriple.of(key, countn, componentMap.build());
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java]: 1. use the patch directly not the map
2. this method checks for `{` for the components but the components use `[` now

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]
Mistakes in the itemstack string format
I suggest using the vanilla path for the item + components section and moving the count to the end

File: Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java (hunk 5 above)
Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java]: I don't know regex 😓

Comment (Mrbysco) [reply Xplat/src/main/java/vazkii/patchouli/common/util/ItemStackUtil.java]: It's probably the `res.indexOf("{")` isn't it 😓

**Comment by Mrbysco** [comment]
Requesting help. the nbt/components aren't actually being read and I have no clue why. 
My IDE doesn't allow me to hotswap with Java 21

File: Xplat/src/main/java/vazkii/patchouli/client/book/template/VariableAssigner.java [hunk 6]
```diff
@@ -169,7 +171,7 @@ private static IVariable inv(IVariable arg) {
 	}
 
 	private static IVariable stacks(IVariable arg) {
-		return IVariable.from(arg.as(Ingredient.class).getItems());
+		return IVariable.from(arg.as(Ingredient.class).getItems(), RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY));
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/client/book/template/VariableAssigner.java]: is this ok? ingredients can use components and those can reference dynamic registries.

File: Xplat/src/main/java/vazkii/patchouli/common/book/Book.java [hunk 7]
```diff
@@ -158,7 +159,7 @@ public Book(JsonObject root, XplatModContainer owner, ResourceLocation id, boole
 		if (noBook) {
 			// Parse on load to catch errors, but need lazy loading for mods
 			// that load after Patchouli
-			var parsed = ItemStackUtil.parseItemStackString(customBookItem);
+			var parsed = ItemStackUtil.deserializeStack(customBookItem, VanillaRegistries.createLookup());
```
Comment (Minecraftschurli) [inline Xplat/src/main/java/vazkii/patchouli/common/book/Book.java:162]: this does an itemstack lookup and also could have dynamic registries in the components (see enchantments in 1.21)

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

File: Xplat/src/main/java/vazkii/patchouli/common/book/Book.java (hunk 7 above)
Comment (Minecraftschurli) [reply Xplat/src/main/java/vazkii/patchouli/common/book/Book.java:162]: Ok I checked and this runs way before there are dynamic registries (which is a problem for another time)

**Review by Minecraftschurli — APPROVED** [review]
